File: src/class068/C4_FillCellsUseAllColorsWays.py
MOD = 1000000007
MAXN = 5001

# 使用二维列表来存储动态规划的结果
dp = [[0] * MAXN for _ in range(MAXN)]

def ways2(n, m):
    # 初始化第一列，只有一种颜色的情况
    for i in range(1, n + 1):
        dp[i][1] = m
    
    # 动态规划填表
    for i in range(2, n + 1):
        for j in range(2, m + 1):
            dp[i][j] = (dp[i-1][j] * j) % MOD
            dp[i][j] = ((dp[i-1][j-1] * (m - j + 1)) + dp[i][j]) % MOD
    
    return dp[n][m]

def main():
    # 功能测试
    N = 9
    M = 9
    print("功能测试开始")
    for n in range(1, N + 1):
        for m in range(1, M + 1):
            result = ways2(n, m)
            print(f"n = {n}, m = {m}, ways = {result}")
    print("功能测试结束")

    # 性能测试
    n = 5000
    m = 4877
    print("性能测试开始")
    print(f"n = {n}, m = {m}")
    import time
    start = time.time()
    ans = ways2(n, m)
    end = time.time()
    print("取余之后的结果 :", ans)
    print("运行时间 :", (end - start) * 1000, "毫秒")
    print("性能测试结束")

if __name__ == "__main__":
    main()

File: src/class068/C4_FillCellsUseAllColorsWays.cpp
#include <iostream>
#include <vector>
using namespace std;

const int MAXN = 5001;
const int mod = 1000000007;
vector<vector<int>> dp(MAXN, vector<int>(MAXN, 0));

// Dynamic programming approach
int ways2(int n, int m) {
    for (int i = 1; i <= n; i++) {
        dp[i][1] = m;
    }
    for (int i = 2; i <= n; i++) {
        for (int j = 2; j <= m; j++) {
            dp[i][j] = (int)(((long long)dp[i - 1][j] * j) % mod);
            dp[i][j] = (int)((((long long)dp[i - 1][j - 1] * (m - j + 1)) + dp[i][j]) % mod);
        }
    }
    return dp[n][m];
}

int main() {
    // Test cases
    int N = 9;
    int M = 9;
    cout << "Functionality test begins" << endl;
    for (int n = 1; n <= N; n++) {
        for (int m = 1; m <= M; m++) {
            int ans = ways2(n, m);
            cout << "n = " << n << ", m = " << m << ", ways = " << ans << endl;
        }
    }
    cout << "Functionality test ends" << endl;

    // Performance test
    int n = 5000;
    int m = 4877;
    cout << "Performance test begins" << endl;
    cout << "n = " << n << ", m = " << m << endl;
    long start = clock();
    int result = ways2(n, m);
    long end = clock();
    cout << "Result after modulo: " << result << endl;
    cout << "Execution time: " << (end - start) / double(CLOCKS_PER_SEC) * 1000 << " milliseconds" << endl;
    cout << "Performance test ends" << endl;

    return 0;
}
